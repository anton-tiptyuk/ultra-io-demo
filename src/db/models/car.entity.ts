import {
  Entity,
  PrimaryColumn,
  Column,
  ManyToOne,
  OneToMany,
} from 'typeorm';

import { decimalTransformer } from '../../common/util';

import { Manufacturer } from './manufacturer.entity';
import { CarToOwner } from './carToOwner.entity';

@Entity()
// @Index('uq_box_title_lower', { synchronize: false })
export class Car {
  @PrimaryColumn({ length: 40 })
  id: string;

  @ManyToOne(() => Manufacturer, { onDelete: 'CASCADE', nullable: false })
  manufacturer: Manufacturer;

  @Column('decimal', { precision: 10, scale: 2, transformer: decimalTransformer })
  price: number;

  @Column()
  firstRegistrationDate: Date;

  @OneToMany(() => CarToOwner, carToOwners => carToOwners.car)
  carToOwners: CarToOwner[];
}
