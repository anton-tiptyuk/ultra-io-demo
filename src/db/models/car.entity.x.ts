import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  OneToMany,
  Index,
  BeforeUpdate,
} from 'typeorm';

// import { Exclude } from 'class-transformer';
// import { BoxToUser } from './boxToUser.entity';

@Entity()
// @Index('uq_box_title_lower', { synchronize: false })
export class Car {
  @PrimaryGeneratedColumn() id: number;

  // @Column({ length: 100 })
  // title: string;

  // @Column({ length: 100, unique: true })
  // barcode: string;

  // @Column({ length: 5 })
  // currency: string;

  // @ManyToOne((type) => PayDestination, { onDelete: 'CASCADE', nullable: true })
  // payDestination: PayDestination;

  // @Column()
  // allowCash: boolean;

  // @Column()
  // allowCashless: boolean;

  // @Exclude()
  // @CreateDateColumn()
  // createdAt: Date;

  // @Exclude()
  // @UpdateDateColumn()
  // updatedAt: Date;

  // @Exclude()
  // @OneToMany((type) => BoxToUser, boxToUser => boxToUser.box)
  // boxToUsers: BoxToUser[];

  // private checkPaymentStuff() {
  //   if (this.payDestination && !this.allowCashless)
  //     throw new BadRequestException('You should enable cashless payments to set payment destination');
  //   if (this.allowCashless && !this.payDestination)
  //     throw new BadRequestException('You should set payment destination to enable cashless payments');
  //   if (!(this.allowCash || this.allowCashless))
  //     throw new BadRequestException('The box should have at least one payment method enabled');
  // }

  // @BeforeUpdate()
  // beforeUpdate() {
  //   this.checkPaymentStuff();
  // }

}
