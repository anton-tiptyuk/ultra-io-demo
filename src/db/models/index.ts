export * from './owner.entity';
export * from './manufacturer.entity';
export * from './car.entity';