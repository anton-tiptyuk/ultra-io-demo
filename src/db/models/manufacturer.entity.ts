import {
  Entity,
  PrimaryColumn,
  Column,
} from 'typeorm';

@Entity()
export class Manufacturer {
  @PrimaryColumn()
  id: string;

  @Column()
  name: string;

  @Column()
  phone: string;

  @Column()
  siret: string;
}
