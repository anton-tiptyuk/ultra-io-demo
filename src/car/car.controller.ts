import { Controller, Get, Param } from '@nestjs/common';
import { Crud, CrudController } from '@nestjsx/crud';

import { Car } from '../db/models';
import { CarService } from './car.service';

@Crud({
  model: { type: Car },
  query: {
    join: {
      manufacturer: { eager: true },
    },
  },
  params: {
    id: {
      // type: 'uuid',
      // though i seed UIDs for all entitites, the task says ID fields should be string
      type: 'string',
      primary: true,
      field: 'id',
    },
  },
  routes: { exclude: ['createManyBase', 'replaceOneBase'] },
})
@Controller('car')
export class CarController implements CrudController<Car> {
  constructor(public service: CarService) { }

  @Get(':id/manufacturer')
  getManufacturer(@Param('id') id: string) {
    return this.service.getManufacturer(id);
  }
}
