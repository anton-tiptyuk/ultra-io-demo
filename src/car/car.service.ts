import { Injectable } from '@nestjs/common';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';

import { Car } from '../db/models';

@Injectable()
export class CarService extends TypeOrmCrudService<Car> {
  constructor(@InjectRepository(Car) repo: Repository<Car>) {
    super(repo);
  }

  decidePagination() {
    return false;
  }
}
