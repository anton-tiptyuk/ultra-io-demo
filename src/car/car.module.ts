import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { CarController } from './car.controller';
import { CarService } from './car.service';
import { Car, CarToOwner } from '../db/models';

@Module({
  imports: [TypeOrmModule.forFeature([Car, CarToOwner])],
  controllers: [CarController],
  providers: [CarService],
  exports: [CarService],
})
export class CarModule { }
