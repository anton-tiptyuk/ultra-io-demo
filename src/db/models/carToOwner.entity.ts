import {
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  Column,
} from 'typeorm';

import { Car } from './car.entity';
import { Owner } from './owner.entity';

@Entity()
export class CarToOwner {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne(() => Car, car => car.carToOwners, { onDelete: 'CASCADE' })
  car: Car;

  @ManyToOne(() => Owner, owner => owner.carToOwners, { onDelete: 'CASCADE' })
  owner: Owner;

  @Column()
  purchaseDate: Date;
}
