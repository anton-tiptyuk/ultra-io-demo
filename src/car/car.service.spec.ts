import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';

import { Car, CarToOwner } from '../db/models';
import { CarService } from './car.service';

const carRepoMock = {
  // Base TypeOrmCrudService<> constructor pacifier
  metadata: { connection: { options: { type: 'anyDbType' } } },
  findOneOrFail: () => undefined,
  update: () => undefined,
};

const carToOwnerRepoMock = {
  delete: () => undefined,
};

const spies = {
  carRepo: {
    findOneOrFail: jest.spyOn(carRepoMock, 'findOneOrFail'),
    update: jest.spyOn(carRepoMock, 'update'),
  },
  carToOwnerRepo: {
    delete: jest.spyOn(carToOwnerRepoMock, 'delete'),
  },
};

let service: CarService;

describe('CarService', () => {
  beforeAll(async () => {
    const blankHandler = () => undefined;
    // Base TypeOrmCrudService<> constructor pacifier
    (TypeOrmCrudService.prototype as any).onInitMapEntityColumns = blankHandler;
    (TypeOrmCrudService.prototype as any).onInitMapRelations = blankHandler;

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        CarService,
        {
          provide: getRepositoryToken(Car),
          useValue: carRepoMock,
        },
        {
          provide: getRepositoryToken(CarToOwner),
          useValue: carToOwnerRepoMock,
        },
      ],
    }).compile();

    service = module.get(CarService);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should disable pagination', () => {
    expect(service.decidePagination()).toBe(false);
  });

  it('should look for manufacturer', async () => {
    const manufacturerId = '9357099e-1c1a-4ff0-8adc-084bf89daf94';
    const manufacturer = {
      id: manufacturerId,
      anyField: 'value'
    };

    spies.carRepo.findOneOrFail.mockImplementation(
      async () => ({ manufacturer })
    );

    await expect(service.getManufacturer(manufacturerId)).resolves.toEqual(manufacturer);
    expect(spies.carRepo.findOneOrFail)
      .toHaveBeenCalledWith(manufacturerId, { relations: ['manufacturer'] });
  });

  it('should perform aging', async () => {
    spies.carRepo.update.mockImplementation(async () => 'update result');
    spies.carToOwnerRepo.delete.mockImplementation(async () => 'delete result');

    await expect(service.performAging()).resolves.toMatchSnapshot();

    const updateCalls = spies.carRepo.update.mock.calls;
    expect(updateCalls.length).toBe(1);
    expect(updateCalls[0].length).toBe(2);
    expect((updateCalls as any)[0][1]['price']()).toBe('price * 0.8');

    expect(spies.carToOwnerRepo.delete.mock.calls.length).toBe(1);
    expect(spies.carToOwnerRepo.delete.mock.calls[0].length).toBe(1);
  });

});
