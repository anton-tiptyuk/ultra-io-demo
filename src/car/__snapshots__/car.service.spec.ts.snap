// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarService should perform aging 1`] = `
Array [
  "delete result",
  "update result",
]
`;
